src/main/java/bio/terra/common/stairway/StairwayComponent.java

```diff
@@ -142,4 +152,63 @@ private String getClusterName() {
     return String.format(
         "%s-%s", kubeService.getNamespace(), stairwayProperties.getClusterNameSuffix());
   }
+
+  public enum Status {
+    INITIALIZING,
+    OK,
+    ERROR,
+    SHUTDOWN,
+  }
+
+  /**
+   * Builder for providing inputs to the initialize method. This allows us to add parameters without
+   * adding new methods to the Stairway component.
+   */
+  public static class InitializerBuilder {
+    private DataSource dataSource;
+    private Object context;
+    private List<StairwayHook> hooks = new ArrayList<>();
+    private ExceptionSerializer exceptionSerializer;
```

Comment:
WDYT about having a `Function<Stairway.Builder, Stairway.Builder>` (or `Consumer<Stairway.Builder>`) to allow clients to mess with the underlying Stairway.Builder anyway they want? That means we don't need to keep extending this object for every new field on Stairway.Builder. Yes, it does allow the client to shoot themselves in the foot by overriding fields `StairwayComponent` sets, but I'm not convinced preventing bad ideas is statically enforceable anyway. We could always set the "StairwayComponent reserved fields" after applying the clients `Function`

We could either make that a field on `InitializerBuilder` or just get rid of the wrapping `InitializerBuilder` and take the `Function` directly in `initialize`. e.g.
```
// StairwayComponent.java
public void initialize(Function<Stairway.Builder, Stairway.Builder> stairwayCustomization);

// client
stairwayComponent.initialize(stairwayBuilder -> {
    return stairwayBuilder
      .applicationContext(myContext)
      .exceptionSerializer(mySerializer);
    // etc.
  });

Reply:
One of the goals of the StairwayComponent was to have a consistent set of stairway configuration parameters that we could share. Sharing the setting of the parameters between StairwayComponent and service code seems confusing and hard to use to me. Yes it is a PITA to extend this builder when a new runtime Stairway parameter is added.

It won't be every time though. For example, the garbage collection controls are all config params and won't show up here. Only runtime things need to be provided on initialize.

I think the exception serializer can become common for us, if/when we move JobControl into the common lib.

In any case, my intent here was to solve this one problem for WSM production, not to redesign the StairwayComponent. If you don't want this introduced, I'll make a new method that passes in the serializer. This seemed more future-proof to me.

Reply:
SG, adding a builder seems better than a new method or continuing to extend the initialize parameters.
The parameter sharing does seem confusing, but agreed on punting a larger redesign.